relay-server/src/actors/project_upstream.rs

```diff
@@ -182,200 +222,261 @@ impl UpstreamProjectSource {
+                            continue;
                         }
+                        let state = response
+                            .configs
+                            .remove(&key)
+                            .unwrap_or(ErrorBoundary::Ok(None))
+                            .unwrap_or_else(|error| {
+                                relay_log::error!(
+                                    "error fetching project state {}: {}",
+                                    key,
+                                    LogError(error)
+                                );
+                                Some(ProjectState::err())
+                            })
+                            .unwrap_or_else(ProjectState::missing);
+                        let result = if state.invalid() { "invalid" } else { "ok" };
+                        metric!(
+                            histogram(RelayHistograms::ProjectStateAttempts) = channel.attempts,
+                            result = result,
+                        );
+                        metric!(
+                            counter(RelayCounters::ProjectUpstreamCompleted) += 1,
+                            result = result,
+                        );
+                        channel.send(state.sanitize());
                     }
                 }
+                Ok((channels_batch, Err(err))) => {
+                    relay_log::error!("error fetching project states: {}", LogError(&err));
 
-                if !slf.state_channels.is_empty() {
-                    // we still have some project configs waiting for state
-                    // try again next time
-                    slf.schedule_fetch(ctx);
-                } else {
-                    // No open channels left, if this is because we fetched everything we
-                    // have already reset the backoff. If however, this is because we had
-                    // failures but the channels have been cleaned up because the requests
-                    // expired we need to reset the backoff so that the next request is not
-                    // simply ignored (by handle) and does a schedule_fetch().
-                    // Explanation 2: We use the backoff member for two purposes:
-                    //  -1 to schedule repeated fetch requests (at less and less frequent intervals)
-                    //  -2 as a flag to know if a fetch is already scheduled.
-                    // Resetting it in here signals that we don't have a backoff scheduled (either
-                    // because everything went fine or because all the requests have expired).
-                    // Next time a user wants a project it should schedule fetch requests.
-                    slf.backoff.reset();
+                    // Put the channels back into the queue, we will retry again shortly.
+                    state_channels.extend(channels_batch);
+                }
+                Err(err) => {
```

Comment:
In this case, should we not also preserve `channels_batch` and put it back into `state_channels`?

Reply:
done f8c83d2